### PR TITLE
[Perf] Make TemplatedParent a direct property.

### PR DESCRIPTION
**Pull request by grokys**
During initial layout, `TemplatedParent` is read tens of thousands of times. This being a styled property that used `inherits: true` meant that it was showing up in profiling as taking up a significant amount of time.

Make `TemplatedParent` a direct property so that reading it is a simple property access. This doesn't actually complicate the code much at all as once set the property value doesn't change, so `inherited: true` semantics are not that important.

I tested this by profiling an initial load of the calendar page of ControlCatalog using dotTrace:

Before:

before

After:

after

Also snuck in a change to make ncrunch ignore Avalonia.Native.

### CODE REVIEW

File: src/Avalonia.Controls/ItemsControl.cs [hunk 1]
```diff
@@ -249,8 +249,6 @@ protected virtual void OnContainersMaterialized(ItemContainerEventArgs e)
                         if (containerControl != null)
                         {
                             ((ISetLogicalParent)containerControl).SetParent(this);
-                            containerControl.SetValue(TemplatedParentProperty, null);
```
Comment (grokys) [inline src/Avalonia.Controls/ItemsControl.cs]: This is no longer needed because it was clearing the inherited value. Now we have no inheritance.

**Review by danwalmsley — APPROVED** [review]
Tested on Windows AS and Mac AS

File: src/Avalonia.Controls/Primitives/TemplatedControl.cs [hunk 2]
```diff
@@ -326,6 +326,23 @@ protected virtual void OnTemplateChanged(AvaloniaPropertyChangedEventArgs e)
             InvalidateMeasure();
         }
 
+        /// <summary>
+        /// Sets the TemplatedParent property for the created template children.
+        /// </summary>
+        /// <param name="control">The control.</param>
+        private void ApplyTemplatedParent(IControl control)
```
Comment (kekekeks) [inline src/Avalonia.Controls/Primitives/TemplatedControl.cs]: What happens when controls are added after ApplyTemplate? What happens if some control is detached? I don't see any code in OnAttached/DetachedToVisualTree that tracks TemplatedParent property of the parent control.

**Comment by kekekeks** [comment]
I think we need to add some kind of "read-optimized" inherited properties that are tracked on property change and attach/detach events.

**Comment by ahopper** [comment]
I measured the ratio of property reads to writes as about 100:1, I think this will drop considerably with the style cache optimization but very much supports the idea of 'read optimized' properties.

**Comment by kekekeks** [comment]
I'm not concerned about write vs. read performance. Having each control to cache inherited property value increases memory consumption by a ton. 
In case of this PR each control now occupies 8 additional bytes of RAM. 

Basically, we can't have each property to behave like that, but we can select some "hot" ones.

**Comment by ahopper** [comment]
ok, In the case of this property it could just be cached by the apply styles code as the same value is used repeatedly as styles are matched to a given control.

**Comment by kekekeks** [comment]
It's OK for TemplatedParent to have a special field, since it's commonly used, so memory tradeoff is justified here.

**Comment by grokys** [comment]
> I think we need to add some kind of "read-optimized" inherited properties that are tracked on property change and attach/detach events.

Yeah, I started work on this and got it working enough to see that it didn't improve performance as much as these other things. I'm sure there will be a time (probably quite soon) where inherited properties need addressing according to profiling. I'll go back to that prototype and finish it off then.

**Pull request merged**